edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/handlers/SubscriptionChangeHandler.cs

```diff
@@ -86,45 +91,211 @@ public async Task<bool> HandleAsync(MqttPublishInfo publishInfo)
+                    {
+                        var subscriptionMatch = Regex.Match(subscription, subscriptionPattern.Pattern);
+                        if (subscriptionMatch.Success)
+                        {
+                            var subscribedDevId = subscriptionMatch.Groups["id1"].Success ? Option.Some<string>(subscriptionMatch.Groups["id1"].Value) : Option.None<string>();
+                            var subscribedModId = subscriptionMatch.Groups["id2"].Success ? Option.Some<string>(subscriptionMatch.Groups["id2"].Value) : Option.None<string>();
+
+                            if (IsMatchingIds(subscribedDevId, subscribedModId, deviceId, moduleId))
+                            {
+                                subscribes = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    try
+                    {
+                        await AddOrRemoveSubscription(listener, subscribes, subscriptionPattern.Subscription);
+                    }
+                    catch (Exception e)
+                    {
+                        Events.FailedToChangeSubscriptionState(e, subscriptionPattern.Subscription.ToString(), identity.Id);
+                    }
+                }
+            }
+            else
+            {
+                Events.CouldNotObtainListener(identity.Id);
+            }
+        }
+
+        // A connection is nested when happens through a lower level edge device. In this case the
+        // current level learns about the connections by 'sidechannels', e.g. when an nested client
+        // subscribes to a topic. The notification of the subscription is assigned to $edgeHub and
+        // and it contains the current subscription list of all nested-connected devices/modules.
+        // There are two issues with this approach:
+        //  1) it can happen that a subscription is sent by a device has never been seen before.
+        //  2) if it is an unsubscribe, it can happen that no device/module related string can be
+        //     found in the subscription list. E.g. if there is a device2 and the subscription list
+        //     is ['device1/twin/res#'], then this notification also means that device2 is unsubscribed
+        //     from all iothub related topics.
+        // To solve 1), the connectionRegistry.GetDeviceListenerAsync() creates the device if it has never
+        // been seen before.
+        // To solve 2), we need to take all known devices and unsubscribe from all topics not listed in the
+        // current notification.
+        async Task HandleNestedSubscriptionChanges(List<string> subscriptionList)
+        {
+            // Collect all added subscriptions to this dictionary for every client. At the end we are going to
+            // unsubscribe from everything not in this collection
+            var affectedClients = new Dictionary<IIdentity, HashSet<DeviceSubscription>>();
```

Comment:
would extract collecting affectedClient in a method

Reply:
extracted